### PR TITLE
Malek/udf grammar changes

### PR DESCRIPTION
**Pull request by MalekiRe**
Changes to match with our new grammar, and also support imperative functions.

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/strings/PowerFxResources.en-US.resx [hunk 1]
```diff
@@ -6534,13 +6534,17 @@
     <comment>Error Message</comment>
   </data>
   <data name="ErrUDF_MissingColon" xml:space="preserve">
-    <value>User Defined Function is missing an ':' here</value>
+    <value>User Defined Function is missing a ':' here</value>
```
Comment (lesaltzm) [inline src/strings/PowerFxResources.en-US.resx:6537]: nit: Probably, `Function definition is missing ...` for these, UDF is the feature name but feels kind of awkward in an error message

File: src/tests/Microsoft.PowerFx.Core.Tests/NamedFormulasTests.cs [hunk 2]
```diff
@@ -28,12 +28,41 @@ public void DefFuncTest(string script)
         }
 
         [Theory]
-        [InlineData("Foo(x As Number, x As String) As Number = Abs(x);")]
-        public void DoubleArgDefTest(string script)
+        [InlineData("Rec4(x: Number): Number => { { force: 1, goo: x } };" +
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/NamedFormulasTests.cs:31]: also some tests with newlines of various kinds, comments between udfs, before/after args, excessive whitespace, etc..

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs [hunk 3]
```diff
@@ -733,26 +739,39 @@ public void IndirectRecursionTest()
...
+        }
+
+        [Fact]
+        public void TestMultiReturn()
+        {
+            var recalcEngine = new RecalcEngine(new PowerFxConfig(null));
+            var str = "Foo(x: Number): Number => { 1+1; 2+2; };";
+            foreach (var error in recalcEngine.DefineFunctions(str).Errors)
+            {
+                Debug.WriteLine($"{error} frag: {error.Span.GetFragment(str)}");
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs:771]: nit: clean up?

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/CheckWrapper.cs [hunk 4]
```diff
@@ -16,14 +16,20 @@ internal class CheckWrapper
...
-        public CheckWrapper(RecalcEngine engine, string expressionText, RecordType parameterType = null)
+        public CheckWrapper(RecalcEngine engine, string expressionText, RecordType parameterType = null, bool isImperative = false)
         {
             _engine = engine;
             _expressionText = expressionText;
             _parameterType = parameterType;
+            ParserOptions = new ParserOptions();
+
+            // If culture isn't explicitly set, use the one from PowerFx Config
+            ParserOptions.Culture ??= _engine.Config.CultureInfo;
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Interpreter/UDF/CheckWrapper.cs:29]: nit: can be an initialization body on ParserOptions:
`
ParserOptions = new ParserOptions() 
{
Culture = _engine.Config.CultureInfo,
AllowsSideEffects = isImperative
}

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 5]
```diff
@@ -1184,7 +1220,7 @@ private TexlNode ParseExprChain(TexlNode node, ITexlSource leftTrivia)
                 sourceList.Add(new TokenSource(delimiter));
                 sourceList.Add(ParseTrivia());
 
-                if (_curs.TidCur == TokKind.Eof || _curs.TidCur == TokKind.Comma || _curs.TidCur == TokKind.ParenClose)
+                if (_curs.TidCur == TokKind.Eof || _curs.TidCur == TokKind.Comma || _curs.TidCur == TokKind.ParenClose || _curs.TidCur == TokKind.CurlyClose)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:1223]: If we're not parsing udfs, does an expression `Foo; }; Bar` parse differently? It'd have been erroneous, so that's safe to change, but double checking.

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 6]
```diff
@@ -837,6 +870,9 @@ private TexlNode ParseOperand()
                     _curs.TokMove();
                     return ParseOperand();
 
+                /*case TokKind.CurlyClose:
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:873]: commented out

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 7]
```diff
@@ -67,113 +69,143 @@ public static ParseUDFsResult ParseUDFsScript(string script, CultureInfo loc = n
...
-                    {
-                        CreateError(_curs.TokCur, TexlStrings.ErrUDF_MissingComma);
-                        return new ParseUDFsResult(udfs, _errors);
-                    }
-                    else if (_curs.TokCur.Kind == TokKind.Comma)
-                    {
-                        TokEat(TokKind.Comma);
-                    }
-                }
+        private bool ParseUDF(List<UDF> udfs)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:112]: probably should be `bool TryParseUdf(out UDF)`, and the caller checks for collisions instead of the parser on line 140. 
basically, it's still valid to parse something that collides

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 8]
```diff
@@ -67,113 +69,143 @@ public static ParseUDFsResult ParseUDFsScript(string script, CultureInfo loc = n
...
+            }
 
-                TokEat(TokKind.ParenClose);
+            var args = new Dictionary<string, UDFArg>();
+            if (TokEat(TokKind.ParenOpen) == null)
+            {
+                return false;
+            }
 
+            while (_curs.TokCur.Kind != TokKind.ParenClose)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:128]: Consider moving ParseUdfArg to it's own function, just for readability.

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 9]
```diff
@@ -67,113 +69,143 @@ public static ParseUDFsResult ParseUDFsScript(string script, CultureInfo loc = n
...
+            {
+                return false;
+            }
+
+            ParseTrivia();
 
-                    // Parse expression
-                    var result = ParseExpr(Precedence.None);
-                    udfs.Add(new UDF(ident.As<IdentToken>(), returnType.As<IdentToken>(), new HashSet<UDFArg>(args.Values), result));
+            var returnType = TokEat(TokKind.Ident);
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:169]: Might be worth allowing TokEat to skip the error, so we can create a more useful one in these cases.

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 10]
```diff
@@ -67,113 +69,143 @@ public static ParseUDFsResult ParseUDFsScript(string script, CultureInfo loc = n
...
                 {
-                    CreateError(_curs.TokCur, TexlStrings.ErrUDF_MissingIdentifier);
-                    break;
+                    return false;
                 }
 
-                ParseTrivia();
-
-                if (TokEat(TokKind.Equ) == null)
+                args.Add(varIdent.ToString(), new UDFArg(varIdent.As<IdentToken>(), varType.As<IdentToken>()));
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:146]: Consider keeping the tokens around, instead of stringifying them, so we could create errors on them later if needed.

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 11]
```diff
@@ -67,113 +69,143 @@ public static ParseUDFsResult ParseUDFsScript(string script, CultureInfo loc = n
...
+            // <bracs-exp> ::= '{' (((<EXP> ';')+ <EXP>) | <EXP>) (';')? '}'
+
+            ParseTrivia();
+
+            if (TokPlainEat(TokKind.CurlyOpen) != null)
+            {
+                _hasSemicolon = false;
+                ParseTrivia();
+                _flags = Flags.EnableExpressionChaining;
+                var result1 = ParseExpr(Precedence.None);
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:192]: nit: better name than result1

File: src/strings/PowerFxResources.en-US.resx (hunk 1 above)
Comment (MalekiRe) [reply src/strings/PowerFxResources.en-US.resx]: Removed these messages entirely.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs (hunk 5 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:1226]: Yes

**Comment by BruceHaley** [comment]
:x: 2 Binary Compatibility issues for **Microsoft.PowerFx.Connectors** compared against [version 0.2.2-preview.20220803-315292](https://www.nuget.org/packages/Microsoft.PowerFx.Connectors/0.2.2-preview.20220803-315292).
<details>

```
MembersMustExist : Member 'Microsoft.PowerFx.Connectors.PowerPlatformConnectorClient.GetAuthToken.get()' does not exist in the implementation but it does exist in the contract.
MembersMustExist : Member 'Microsoft.PowerFx.Connectors.PowerPlatformConnectorClient.Transform(System.Net.Http.HttpRequestMessage)' does not exist in the implementation but it does exist in the contract.
```

</details>
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 12]
```diff
@@ -27,8 +27,10 @@ public enum Flags
             NamedFormulas = 1 << 1
         }
 
+        private bool _hasSemicolon = false;
+
         private readonly TokenCursor _curs;
-        private readonly Flags _flags;
+        private Flags _flags;
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:33]: This being mutable is a "red flag". Should we be using a stack here instead?

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 13]
```diff
@@ -67,113 +69,153 @@ public static ParseUDFsResult ParseUDFsScript(string script, CultureInfo loc = n
...
-                    else if (_curs.TokCur.Kind == TokKind.Comma)
-                    {
-                        TokEat(TokKind.Comma);
-                    }
-                }
+            return new ParseUDFsResult(udfs, _errors);
+        }
 
-                TokEat(TokKind.ParenClose);
+        private bool ParseUDFArgs(HashSet<UDFArg> args)
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:112]: Make this a ref variable. I know it's not required for mutable collections, but it helps the user know that the collection is being modified.

In fact, the new hash set can be created inside this function, and the param marked as "out" instead of "ref".

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 14]
```diff
@@ -67,113 +69,153 @@ public static ParseUDFsResult ParseUDFsScript(string script, CultureInfo loc = n
             return parser.ParseUDFs(script);
         }
 
+        /// <summary>
+        /// If the token is of the right type, it returns it and moves forward.
+        /// This does not post an error if the tok is of the wrong type.
+        /// </summary>
+        /// <param name="tid"></param>
+        /// <returns></returns>
+        private Token TokPlainEat(TokKind tid)
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:78]: Do we need this function? There are existing ways of checking to see if the next token is of a specific type.

**Review by bryancroteau-MSFT — CHANGES_REQUESTED** [review]

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs (hunk 12 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs]: Changed

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs (hunk 14 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs]: Removed.

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs (hunk 13 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs]: Changed

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 15]
```diff
@@ -71,109 +76,136 @@ private ParseUDFsResult ParseUDFs(string script)
...
                 }
 
-                TokEat(TokKind.Semicolon);
+                udfs.Add(new UDF(ident.As<IdentToken>(), returnType.As<IdentToken>(), new HashSet<UDFArg>(args), exp_result, _hasSemicolon));
+                _flagsMode.Pop();
+                return true;
             }
 
-            return new ParseUDFsResult(udfs, _errors);
+            _flagsMode.Push(Flags.NamedFormulas);
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:202]: This is incorrect, you are pushing onto the stack 2 times

**Review by bryancroteau-MSFT — CHANGES_REQUESTED** [review]

**Comment by BruceHaley** [comment]
:x: 2 Binary Compatibility issues for **Microsoft.PowerFx.Connectors** compared against [version 0.2.2-preview.20220803-315292](https://www.nuget.org/packages/Microsoft.PowerFx.Connectors/0.2.2-preview.20220803-315292).
<details>

```
MembersMustExist : Member 'Microsoft.PowerFx.Connectors.PowerPlatformConnectorClient.GetAuthToken.get()' does not exist in the implementation but it does exist in the contract.
MembersMustExist : Member 'Microsoft.PowerFx.Connectors.PowerPlatformConnectorClient.Transform(System.Net.Http.HttpRequestMessage)' does not exist in the implementation but it does exist in the contract.
```

</details>
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs (hunk 15 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs]: It isn't, however the second push and pop group is unnecessary as it is simply is the same as the top level flags.

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 16]
```diff
@@ -53,7 +55,10 @@ private TexlParser(IReadOnlyList<Token> tokens, Flags flags)
 
             _depth = 0;
             _curs = new TokenCursor(tokens);
-            _flags = flags;
+            _flagsMode = new Stack<Flags>();
+            _flagsMode.Push(flags);
+
+            //_flags = flags;
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:61]: remove comment

**Comment by BruceHaley** [comment]
:x: 2 Binary Compatibility issues for **Microsoft.PowerFx.Connectors** compared against [version 0.2.2-preview.20220803-315292](https://www.nuget.org/packages/Microsoft.PowerFx.Connectors/0.2.2-preview.20220803-315292).
<details>

```
MembersMustExist : Member 'Microsoft.PowerFx.Connectors.PowerPlatformConnectorClient.GetAuthToken.get()' does not exist in the implementation but it does exist in the contract.
MembersMustExist : Member 'Microsoft.PowerFx.Connectors.PowerPlatformConnectorClient.Transform(System.Net.Http.HttpRequestMessage)' does not exist in the implementation but it does exist in the contract.
```

</details>
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs [hunk 17]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (bryancroteau-MSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs:6]: Why is this line needed?

**Review by bryancroteau-MSFT — CHANGES_REQUESTED** [review]

**Review by bryancroteau-MSFT — APPROVED** [review]

**Review by bryancroteau-MSFT — APPROVED** [review]

**Comment by BruceHaley** [comment]
:x: 2 Binary Compatibility issues for **Microsoft.PowerFx.Connectors** compared against [version 0.2.2-preview.20220803-315292](https://www.nuget.org/packages/Microsoft.PowerFx.Connectors/0.2.2-preview.20220803-315292).
<details>

```
MembersMustExist : Member 'Microsoft.PowerFx.Connectors.PowerPlatformConnectorClient.GetAuthToken.get()' does not exist in the implementation but it does exist in the contract.
MembersMustExist : Member 'Microsoft.PowerFx.Connectors.PowerPlatformConnectorClient.Transform(System.Net.Http.HttpRequestMessage)' does not exist in the implementation but it does exist in the contract.
```

</details>
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**
